### ISSUE TITLE
All gelbooru sites incorrectly report as down

### ISSUE THREAD

**Issue by GiovanH** [Bionus/imgbrd-grabber#2176]
**Bug description**

In search results, all *.booru.org sites report as 
```
*.booru.org - No Result
Possible reasons: server offline
No valid source of the site returned result
```

But the sites are clearly not actually down when viewed in a browser.

**Expected behavior** Standard functionality

**Context**

Log file looks like:

```
[19:21:39.002][Info] Loading results...
[19:21:39.003][Info] [undertale.booru.org][Html] Loading page https://undertale.booru.org/index.php?page=post&s=list&tags=Undertale&pid=0
[19:21:41.052][Info] [undertale.booru.org][Html] Receiving page https://undertale.booru.org/index.php?page=post&s=list&tags=Undertale&pid=0
[19:21:41.053][Error] [undertale.booru.org][Html] Loading error: Connection closed (2)
[19:21:41.054][Warning] [undertale.booru.org] No valid source of the site returned result.
```

To confirm that this is a grabber issue and not an outage, you can easily navigate to https://undertale.booru.org/index.php?page=post&s=list&tags=Undertale&pid=0 and see that the page is being served correctly.

**System information**

- OS: Windows 10
- Grabber version: 7.3.2 (x64)

**Comment by byte-warlock** [comment]
Having the same issue. It seems to be Gelbooru 0.1.* sources specifically?

**Comment by Smackdab69** [comment]
same issue here

**Comment by GaaraMad** [comment]
If I understood correctly, then the problem is that a cloudflare has appeared on the booru sites and it interferes with the work of the parser

**Comment by FuzzyWuggzy** [comment]
It's not just Gelbooru 0.1 sites, furry.booru.org (Gelbooru 0.2) is reporting 403 despite the html/xml links in the log working fine in browser, tested in v7.2.1 and latest nightly Linux build, no difference, all other Danbooru and Gelbooru sites in my sources are working fine. Cloudflare DDoS protection started popping up on it recently so I'd say GaaraMad's suspicion is correct.

**Comment by GiovanH** [comment]
Can confirm that this is still an issue as of 7.3.2 `fa0a6a00`.

**Comment by 15067** [comment]
I found that changing the user agent value for *.booru.org sites in the headers settings fixes the issue, haven't tested on gelbooru 0.2 sites though.

**Comment by GaaraMad** [comment]
Hm. I added User-Agent, Accept headers and __cfduid, cf_clearance cookies. Didn't work (get 403 error), gelbooru 0.2
Tested on ffe63f07 and fa0a6a00

**Comment by GiovanH** [comment]
> I found that changing the user agent value for *.booru.org sites in the headers settings fixes the issue, haven't tested on gelbooru 0.2 sites though.

I'm also finding that setting a manual header is a workaround, but that's obviously not a good solution for a long list of sites.

According to https://github.com/Bionus/imgbrd-grabber/blob/310089584b5e7c0f62fdc1d4142de21d46635ccb/src/lib/src/models/site.cpp#L280, the user agent header is passed as Grabber. 

One solution would be to add a header override to the site `model.js` as a kind of auth field, but that doesn't seem to currently exist.

**Comment by GiovanH** [comment]
You can temporarily hotpatch the model.js files with these: https://gist.github.com/GiovanH/7ab8fb009e4233c939abb95a05096ad9

**Comment by byte-warlock** [comment]
> You can temporarily hotpatch the model.js files with these: https://gist.github.com/GiovanH/7ab8fb009e4233c939abb95a05096ad9

This partially works for me, I can view the search results now but when I open a post I only see the thumbnail and I'm unable to save the image.

Log output just says the connection was closed:
`
[04:22:11.501][Error] Loading error for 'https://gravityfalls.booru.org/index.php?page=post&s=view&id=58078': Connection closed
`

I'm using version 7.3.2.

**Comment by FuzzyWuggzy** [comment]
> You can temporarily hotpatch the model.js files with these: https://gist.github.com/GiovanH/7ab8fb009e4233c939abb95a05096ad9

Added these to a fresh copy of Grabber 7.3.2 08809d8 that I just cloned and compiled under Arch, still no difference.

I'll try it with the stable Windows release under Wine and report my findings.

**Comment by byte-warlock** [comment]
I added the `User-Agent` header through the source settings in the UI instead of in the `model.js` file and it seems to work.

**Comment by GiovanH** [comment]
> I added the `User-Agent` header through the source settings in the UI instead of in the `model.js` file and it seems to work.

To be clear, all my `model.js` tries to do is add a user-agent header to sites based on the version they're running, since, as I said,

> I'm also finding that setting a manual header is a workaround, but that's obviously not a good solution for a long list of sites.

**Comment by omegarunner1** [comment]
I think I've figured out the problem and there isn't an easy answer, unfortunately. The problem is that, while grabber is downloading en masse, the Gelbooru websites are registering it as a DDOS attack and is blocking your IP address subsequently. The best solution for overcoming this is to get a VPN and just change servers so that you can get a different VPN. Unless your Router comes with a rolling IP address, at which point just wait for it to reset your IP address and you'll be able to use Grabber again.

**Comment by GiovanH** [comment]
No, this happens on all requests including the initial request, not after continued usage.

On Feb 28, 2021, at 10:46 PM, omegarunner1 <notifications@github.com<mailto:notifications@github.com>> wrote:


I think I've figured out the problem and there isn't an easy answer, unfortunately. The problem is that, while grabber is downloading en masse, the Gelbooru websites are registering it as a DDOS attack and is blocking your IP address subsequently. The best solution for overcoming this is to get a VPN and just change servers so that you can get a different VPN. Unless your Router comes with a rolling IP address, at which point just wait for it to reset your IP address and you'll be able to use Grabber again.

**Comment by omegarunner1** [comment]
That's strange because changing the server and IP address in my VPN fixed the issue for me.

**Comment by VanTuz5** [comment]
Same for me. I tried set user-agent and cf_clearance for furry.booru.org, but it didn't work.

**Comment by GiovanH** [comment]
Issue persists in 7.5.1. Seems critical.

**Comment by FlarosOverfield** [comment]
Still happening on 7.10...
